src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+# 
+
+using LinearAlgebra
+ξd(p, n_air, n_metal)= 1 / (n_air + (n_metal - n_air) * p)^2 - 1 / n_air^2 # in the design region
+
+a_base(u, v; phys_params) = (1 / phys_params.n_air^2) * ((∇ .* (Λ(phys_params) * v)) ⊙ (Λ(phys_params) .* ∇(u))) - (phys_params.k^2 * phys_params.μ * (v * u))
+
+a_design(u, v, pth; phys_params) = ((p -> ξd(p, phys_params.n_air, phys_params.n_metal)) ∘ pth) * (∇(v) ⊙ ∇(u))
+
+function MatrixA(pth; phys_params, fem_params)
+    A_mat = assemble_matrix(fem_params.U, fem_params.V) do u, v
+        ∫(a_base(u, v; phys_params))fem_params.dΩ + ∫(a_design(u, v, pth; phys_params))fem_params.dΩ_d
+    end
+    return lu(A_mat)
+end
+
+# ## Solve for plane wave incident
+# 
+# The plane wave source `b_vec` can be simply assembled by a uniform integral over the source line, and the magnetic field vector `u_vec` can then be solved by a simple linear equation
+# 
+
+p0 = zeros(fem_params.np)  # Here we make p=0 everywhere just for illustration purpose
+pf_vec = Filter(p0;r, fem_params)
+pfh = FEFunction(fem_params.Pf, pf_vec)
+pth = (pf -> Threshold(pf; β, η)) ∘ pfh
+A_mat = MatrixA(pth; phys_params, fem_params) 
+b_vec = assemble_vector(v->(∫(v)fem_params.dΓ_s), fem_params.V)
+u_vec = A_mat \ b_vec
+uh = FEFunction(fem_params.U, u_vec)
+
+# ## Objective
+# 
+# The problem is maximizing the electric field intensity at the center. Recall that the electric field can be retrieved from the magnetic field by 
+#
+# ```math
+# \mathbf{E}(\mathbf{x})=\frac{\mathrm{i}}{\omega\varepsilon(\mathbf{x})}\nabla\times\mathbf{H}(\mathbf{x}),
+# ``` 
+# and our objective is the field intensity at center $\vert\mathbf{E}(\mathbf{x}_0)\vert^2$. 
+# 
+# In the 2D formulation, this objective can be simplified to  
+# ```math
+# g = \int \vert\nabla H\vert^2\delta(x-x_0)\mathrm{d}\Omega = u^\dagger Ou,
+# ```
+# where $u$ is the magnetic field vector and  
+# ```math
+# O = \int (\nabla \hat{v}\cdot\nabla\hat{u})\delta(x-x_0)\mathrm{d}\Omega,
+# ```
+# with $\hat{v}$ and $\hat{u}$ are the finite element basis functions. 
+# 
+# In practice, the delta function can be approximated by a concentrated Gaussian function. Note that we use `dΩ_c` here in order to reduce computation costs. 
+# 
+
+function MatrixOf(fem_params)
+    x0 = VectorValue(0,300)  # Position of the field to be optimized
+    δ = 1                    
+    return assemble_matrix(fem_params.U, fem_params.V) do u, v
+        ∫((x->(1/(2*π)*exp(-norm(x - x0)^2 / 2 / δ^2))) * (∇(u) ⋅ ∇(v)) )fem_params.dΩ_c
+    end
+end
+
+# ## Optimization with adjoint method
+# 
+# Now that we have our objective to optimize, the next step is to find out the derivative to the desigin parameter $p$ in order to apply a gradient-based optimization algorithm. We will be using `ChainRulesCore` and `Zygote` packages.  
+# 
+
+using ChainRulesCore, Zygote
+import ChainRulesCore: rrule
```

Comment:
Instead than `importing` the `rrule` function like this, it's usually clearer to simply define your methods as:
```jl
function ChainRulesCore.rrule(...)
   ...
end
```
That way, it's clear at the site of the method definition that it is extending a function from `ChainRulesCore`.